### DISCUSSION TITLE
Inviscid flow an free slip BC

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#23113] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

I am trying to run inviscid flow simulations, to later use as an initial condition for the viscous case in the FE INS module.

However, I have not found a way to implement a free slip/no penetration boundary condition for my geometry - in this case, an aerofoil.

I had thought that a Neumann BC of zero for all velocity components at the boundary would do the trick, and attempted to implement this using 'ADVectorFunctionBC'. 
However, the solution for this was the same as my initial condition (constant velocity for the whole domain), with the flow not moving around the obstacle. The outlet boundary condition was constant pressure = 0.

I came across this thread when looking for a solution: https://github.com/idaholab/moose/discussions/21418
However, I am having trouble understanding the difference between the 'no flux' and 'no penetration' cases described here, as well where the mentioned approaches are implemented (ie - can this be done through a moose input file, or would custom BCs be required?)

My question then, is: Have I misunderstood the relationship between Neumann BCs and the no penetration condition, and what is the proper way to implement the free slip condition for inviscid flows in FE, rather than FV?

Thanks!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Are the flow variables defined inside the aerofoil as well?
In other words, is the aerofoil boundary inside the flow domain or well on the boundary?

Which variable set are you using (velocities or momentum?) ? For whether we have it or not

Guillaume

**Reply by Eilloo** [reply]
The flow variables are not defined in the aerofoil, no. Does this mean that not specifying any boundary condition is the same as what I've manually implemented by setting a Neumann BC to zero?

I'm not certain what you mean by 'variable set', so to be clear, I'm solving for velocity (as well as pressure), which is a Lagrange vector type, but the kernels are 'INSADMomentumAdvection', 'INSADMomentumViscous', and so on.

Thanks!

**Reply by GiudGiud** [reply]
Whether having no BC is equivalent to a 'no-penetration' depends on the equation. It is often equivalent to 'no flux' to have a 0 contribution to the residual at the boundary for balance equations. 

For INSAD, the no-BC or no-penetration boundary condition is the INSADMomentumNoBCBC. 

I asked about the variable set because the flow equations can be solved with P, v, T or with the conservative variable set rho, momentum, energy,

**Reply by Eilloo** [reply]
Okay... I think I'm having trouble understanding what the 'no BC' BC is doing to achieve no penetration. 
The description to me sounds like it is reducing the error compared to a Neumann BC of zero when dealing with outflow boundaries, whereas I don't want any flow passing through my boundary at all.

In terms of the velocity variable, I believe I'm after **u** dot **n** = 0.
Where **u** is the velocity vector and **n** is normal to the boundary.

I think I've convinced myself that this can't be achieved using Neumann BC's imposed on the velocity variable, since this would still allow constant flow through the boundary. Previously I was visualising mass flux, hence my misunderstanding.

Is INSADMomentumNoBCBC still appropriate to impose the above condition?
(As a side note if so, is there a convenient way to pass in components of my velocity variable, since the BC does not accept vectors? If not, I can tweak the code to do so.)

Hopefully what I want to achieve makes sense - thanks again for your help!

**Reply by GiudGiud** [reply]
> For INSAD, the no-BC or no-penetration boundary condition is the INSADMomentumNoBCBC.

Maybe I m wrong here. 

@lindsayad what is the noBCBC

**Reply by lindsayad** [reply]
The convection term is not integrated by parts in the FE implementation (at least currently) and so any integrated boundary condition you set will have no impact on the convection "flux"

**Reply by lindsayad** [reply]
We don't currently have a great slip boundary condition implementation for the incompressible FE formulation. However, I do think there will be one incoming with https://github.com/idaholab/moose/pull/22671 

@GiudGiud the NoBCBC just weakly adds the stresses on the boundaries, so it's basically setting: stress = stress ... so it's doing nothing. Practically speaking it imposes some condition on some higher derivative. Here's what's in our documentation:

> The `INSADMomentumNoBCBC` class is used to implement the "No Boundary Condition" developed by [Griffiths](https://onlinelibrary.wiley.com/doi/abs/10.1002/(SICI)1097-0363(19970228)24:4%3C393::AID-FLD505%3E3.0.CO;2-O). According to Griffiths, the "No Boundary Condition" is equivalent to imposing the condition on the original analytical problem that the (p+1)st derivative of a finite element variable of order p should vanish at a point close to the outflow. This is claimed to reduce error relative to a natural boundary condition.

**Reply by GiudGiud** [reply]
Isnt a 'natural' boundary condition a slip one though? I feel like the 'unnatural', eg the one that you have to do more work for, is setting velocity to 0 for no-slip ?

**Reply by lindsayad** [reply]
The natural boundary condition will set stresses to 0 and that's it. You are used to FV where we convert the advection term into a face term. In the current FE implementation we don't integrate the advection term by parts and so it doesn't appear in surface terms, e.g. integrated boundary conditions

**Reply by Eilloo** [reply]
Thanks both for the clarification, and great to hear there's something on the way!

**Reply by doquang** [reply]
Hi, I would like to know if there is any further update on the free slip BC? Looks like `MomentumFreeSlipBC` was implemented for 2D only. Is there a similar method for 3D?

**Reply by lindsayad** [reply]
You could use the `ConservativeAdvection` or `ADConservativeAdvection` kernel introduced in https://github.com/idaholab/moose/pull/23986 and if you apply it to the mass equation, then a natural boundary condition for the mass equation would be no normal flow through a boundary. An additional natural boundary condition for the momentum equation would impose no normal stress.

**Reply by doquang** [reply]
I am currently using INSAD module for solving the NS equations. Can the Hybrid CG DG `ConservativeAdvection` be applied on this model?

**Reply by lindsayad** [reply]
No because `INSADMomentumAdvection` is not integrated by parts. However, the mass advection is integrated by parts in the `INSFEFluidMassKernel` and there is an option to integrate the momentum advection by parts in `INSFEFluidMomentumKernel`. Those kernels also use continuous FEM but use scalar field components for the velocity and do not compute their Jacobians with AD.

If you used that kernel set, then you could impose free slip boundaries using natural boundary conditions.

**Reply by doquang** [reply]
@lindsayad, Can you help me understand the `expansion-channel-slip-wall.i` test in Navier-Stokes module. Why do we still need to implicitly set the normal component of velocity to 0 (INSFEFluidWallMomentumBC/vy_wall and INFEMomentumFreeSlipBC/slipwall). Even though we already used a conservative form and integration-by-part?

**Reply by lindsayad** [reply]
Hi @doquang it's a good question. The natural condition on the walls for the mass equation do implicitly accomplish the zero normal velocity condition. `INSFEFluidWallMomentumBC` is perfectly valid to add as it allows for non-zero normal stresses. If you did not impose that BC, then you would be imposing these additional conditions at the walls through the natural condition:

$p \vec{n} - \mu \nabla \vec{v} \cdot \vec{n} = \vec{0}$

Now I don't know what the purpose of `INSFVMomentumFreeSlipBC` is in this case. I commented out that boundary condition and ran the simulation and to me the results look good. Maybe @lingzou could weigh in on this

**Reply by lingzou** [reply]
For this case, with or without the `INSFVMomentumFreeSlipBC` look almost the same. In some cases I have seen before, WITHOUT the free slip BC, velocity on a wall would NOT aligned with the wall perfectly so I used that BC to 'help' align the velocity with wall. (unfortunately I now cannot find that case but it did exist 🙄).

If in your case, the velocity alignment is good (or acceptable), I'd suggest just to ignore the `INSFVMomentumFreeSlipBC`.

**Reply by doquang** [reply]
Thanks @lingzou and @lindsayad. BTW, do you have any documentation to learn about how **INSFE** is implemented. For example, why does `INSFEFluidMassKernel` add the advective term of mass conservation equation along with pressure-stabilized Petrov-Galerkin terms. I mean `masseq_part`. Is this not the same as the implementation in the **INS** and **INSAD** modules, right?

**Reply by lingzou** [reply]
Section 9 of SAM theory manual (https://publications.anl.gov/anlpubs/2021/05/166771.pdf), or more specifically, equation (9.29), should answer your question.

[sorry that the codes have been merged but the documentation part we haven't finished.]

**Reply by lindsayad** [reply]
It is the same apart from the fact that for `INSMass` and for `INSADMass` we do not integrate the $\psi \nabla \cdot \vec{v}$ term by parts. You must include $\psi \nabla \cdot \vec{v}$, whether integrated by parts or not, because that is the physics, it is the math enforcing conservation of mass. PSPG is just a stabilization term you add on top of it
